### PR TITLE
OPSIM-1136: Update zero point for calc_season and make it multi-RA capable

### PR DESCRIPTION
**Pull request by rhiannonlynne**
While almost all of the time calc_season would be called with only a single RA, if it's confusing, we can make it multi-RA like this.

More importantly though, it does update the "zero point" for the seasons, to introduce the discontinuity in season count that exists in the scheduler code ..  This PR makes calc_season start counting at 1, and counting 'season' as starting when the RA value matches the RA of the Sun.  

I think this still offsets the zero point for the seasons by about 90 degrees from what the scheduler counts as the "start" of the season though.  I'm not sure if this is going to remain confusing or not.

```
import numpy as np
import matplotlib.pyplot as plt
from rubin_scheduler.utils import survey_start_mjd, calc_season

ra = np.arange(0, 360, 1)
mjds = np.arange(survey_start_mjd(), survey_start_mjd()+11*365.25, 1)
seasons = calc_season(ra, mjds)
seasons = np.swapaxes(seasons, 0, 1)
seasons = np.floor(seasons)

plt.imshow(seasons, aspect=.1, origin='lower')
plt.colorbar()
plt.xlabel("RA (deg)")
plt.ylabel("MJD")
```

Screenshot 2024-03-08 at 7 51 01 PM

### CODE REVIEW

**Comment by rhiannonlynne** [comment]
Oh, interesting. I thought from the comments that Matt Becker made on the rolling cadence pause threads, that the scheduler didn't use calc_season. However, that's not exactly true because "optimize_ddf_times" uses calc_season. 

So now maybe the question is: 
(a) I'm maybe surprised that we never ran into season 11 for the DDFs before?  (because in terms of visibility seasons, some parts of the sky do have 11 seasons, although those are only partial seasons near the end or start of the survey). I don't think we did, because the "season_vals" only have keys 0-10, but I don't know why. 
(b) we should now think about what to do about the -1 season, which is what is tripping up the script. The "-1" season represents a partial season in the first year of the survey, where the DDF wouldn't get the full season to be observed.  With the current survey start date, all of the DDFs have a short season to start the survey, although COSMOS has a slightly longer first partial season.

**Comment by rhiannonlynne** [comment]
Can you tell I'm procrastinating on writing a talk that I am giving on Monday? 

I went and poked at the place where the new season calculation was making the ddf field optimization fail. The "season_val" dictionary in "ddf_slopes" was failing because now there was a "-1" season, which will be the first incomplete season. 

I thought this alternate (see new ddf_slopes) which scales "ss" by the actual season length in any season would be a reasonable alternative, and more truly flexible with survey start date. It should be returning something very similar to, although not the same as, the previous ddf_slope.  The first and last years are sometimes different, but I think this is probably appropriate rather than just automatically giving first or last seasons (which may vary in length) the same default of "10 sequences". 

You can test out the new ddf_slopes compared to the old ddf_slopes by looking at this notebook here - https://github.com/rhiannonlynne/notebooks/blob/main/ddf_slopes_tests.ipynb

The results are similar, but not identical between the two methods of calculating "slope". 
I suspect the new way is better, but another option to restore the old default would be to just subtract the minimum season from the season values after using calc_season, to get back to the old "first season at all is 0" behavior.

**Comment by rhiannonlynne** [comment]
Another place turned up an error -- in the feature "NObservationsCurrentSeason", the calc_season utility is called .. and was previously being called with multiple RA values (at a single time). 
I think this probably means the feature did not work correctly previously .. no matter what time you called calc_season with, because it always got the same range of RA values and only a single time, the resulting "season map" would look the same (the same zero point would be being subtracted every time).

The failure happening in the unit test at the moment is because when setting up the season map, a series of RA values are passed to calc_season (at a single time). 
calc_season called this way will now return a 2-d array shaped like (len(ra), len(mjd)) instead of the previous (len(mjd)). 

I'm not entirely sure what to do to update this feature, because I'm not quite clear what "season_map" is supposed to do. 
I get the idea that the feature is intended to track how many observations have happened in a given season, but I don't understand why it's necessary to have "season_update" .. it might be there partly because of the previous mismatch in what was being returned by calc_season? 

If the point was to track the number of observations that match criteria, I would have thought you have two ra/dec maps contained in the feature -- one containing the number of visits in the current season at that healpix and a second containing "what is the current season being tracked at this ra/dec" .  
When a visit is added to the feature, I would imagine you'd evaluate if the visit meets criteria and then calculate the healpix ids that overlap the visit. Then call calc_season for those healpixes, at this time. For each healpix, check if the new observation is within the same season as the "current_season" map -- if it is the same, just increment the "number of visits in the current season", and if not, reset the "number of visits in the current season" to 1. 

I don't know why the current feature has a line of code that sets the entire sky's "count of visits in the current season" to 0, if the season changed anywhere.

**Comment by rhiannonlynne** [comment]
I suppose another thing to do would be to understand the difference between calc_season (now) and season_calc .. I *think* they should be returning the same value, but modified by 90 degrees (which might mean that doing floor or round puts visits into a different visibility season between the two?).

**Review by yoachim — APPROVED** [review]
Not sure why unit tests passed on git when they failed locally for me. 

We might want to copy over the basic auxtel scheduler setup to add to the unit tests, then we have a check that we're not deprecating anything that's in use.

File: rubin_scheduler/scheduler/basis_functions/basis_functions.py [hunk 1]
```diff
@@ -724,30 +766,46 @@ def __init__(
...
-        offset=None,
+        mjd_start=None,
         season_frac_start=0.5,
     ):
-        super(SeasonCoverageBasisFunction, self).__init__(nside=nside, filtername=filtername)
+        super().__init__(nside=nside, filtername=filtername)
 
-        self.n_per_season = n_per_season
+        if footprint is None:
+            footprint = np.ones(hp.nside2npix(self.nside), float)
```
[the remark this answers is not in the document]
Comment (rhiannonlynne) [reply rubin_scheduler/scheduler/basis_functions/basis_functions.py:775]: Yes .. I can update it (I wrote the np.ones before I added the get_current_footprint  function.

**Comment by rhiannonlynne** [comment]
> Not sure why unit tests passed on git when they failed locally for me.
> 


Ha, actually - the unit tests now fail on GitHub (after the `remove season_ things` commit) .. looks like maybe it's a python environment thing. Did you just update your environment, maybe? (kind of looks like we were implicitly using a package one of our dependencies imported, and now maybe they've stopped .. but will check).

**Comment by rhiannonlynne** [comment]
> We might want to copy over the basic auxtel scheduler setup to add to the unit tests, then we have a check that we're not deprecating anything that's in use.

I'll think about this. I actually don't think just copying the auxtel setup over by itself would keep us up to date, because the auxtel setup can (and does) change. 
But perhaps there could be an extra step somewhere to install ts_fbs_utils and ts_config_ocs (which uses ts_fbs_utils), but it's a little awkward (because while ts_fbs_utils will pip install, ts_config_ocs is not as straightforward).

**Pull request merged**